src/osd/ScrubResult.h

```diff
@@ -0,0 +1,153 @@
+#include "common/map_cacher.hpp"
+
+namespace Scrub {
+
+  /**
+   * errors found when comparing shard with auth replica
+   */
+  struct ShardError {
+    ShardError(const pg_shard_t& shard,
+	       const ScrubMap::object *object = nullptr)
+      : shard(shard), object(object)
+    {}
+    void on_shard_missing();
+    void on_stat_error();
+    void on_read_error();
+    void on_omap_digest_mismatch();
+    void on_data_digest_mismatch();
+    void on_size_mismatch();
+    void on_attr_missing();
+    void on_attr_mismatch();
+    void on_attr_unexpected();
+    void on_attr_corrupted();
+
+    const pg_shard_t shard;
+    const ScrubMap::object *object;
+    uint64_t errors = 0;
+  };
+
+  /**
+   * errors found when 
+   */
+  struct ObjectError {
+    const hobject_t oid;
+    // osd => ScrubMap::object
+    std::map<int32_t, const ScrubMap::object*> shards;
+    uint64_t errors;
+    ObjectError(const hobject_t& oid);
+    void on_auth_missing(const map<pg_shard_t, ScrubMap*>& maps);
+    void on_shard_error(const ShardError& e);
+    // technically, these are ShardErrors, but we don't have
+    // their keys in omap when we encounter them in ReplicatedPG::_scrub().
+    // if we can use the object_t for the key, the problem is solved.
+    //
+    // auth copy errors, most likely SS_ATTR/OI_ATTR
+    void on_attr_missing(const char* name);
+    void on_attr_decode_failure(const char* name);
+    void on_oi_size_mismatch(uint64_t disk_size);
+
+    void encode(bufferlist& bl) const;
+  };
+
+  struct SnapError {
+    uint64_t errors = 0;
+    hobject_t head;
+    bufferlist snapset;
+    vector<snapid_t> clones;
+    SnapError() = default;
+    SnapError(const hobject_t& head);
+    void set_snapset(bufferptr ss_attr);
+    void on_ss_attr_missing();
+    void on_ss_attr_corrupted();
+    // snapset with missing clone
+    void on_clone_missing(snapid_t);
+    // the snapset is not consistent with itself
+    void on_snapset_mismatch();
+    // soid.snap inconsistent with snapset
+    void on_head_mismatch();
+    void encode(bufferlist& bl) const;
+  };
+
+  class Store {
+  public:
+    Store(const coll_t& coll, const hobject_t& oid, ObjectStore* store);
+    ~Store();
+    void add_object_error(const ObjectError& e);
+    void add_snap_error(const SnapError& e);
+    bool empty() const;
+    ObjectStore::Transaction *get_transaction();
+
+  private:
+    ObjectStore::Transaction* _transaction();
+
+  private:
+    // a temp object holding mappings from seq-id to inconsistencies found in
+    // scrubbing
+    OSDriver driver;
+    MapCacher::MapCacher<std::string, bufferlist> backend;
+    ObjectStore::Transaction *txn = nullptr;
+  };
+
+
+  enum SnapErr {
+    CLONE_MISSING = 1 << 1,
+    SNAP_MISMATCH = 1 << 2,
+    ATTR_MISSING  = 1 << 3,
+    HEAD_MISMATCH = 1 << 4,
+    SIZE_MISMATCH = 1 << 5,
+  };
+
+  inline hobject_t make_scrub_object(const pg_t& pgid)
+  {
+    ostringstream ss;
+    ss << "scrub_" << pgid;
```

Comment:
@dachary the scrub objects are stored in the temp collection of the scrubbed PG. and they are created as temp objects. AFAIK, user is not able to access the temp collection or the temp pool, and i also reviewed the code by searching the callers of `coll_t::get_temp()`, seems i am the only user of the temp collection =)